internal/warn_deprecated_api.ts

```diff
@@ -0,0 +1,47 @@
+// Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.
+// This module is browser compatible.
+
+// deno-lint-ignore no-explicit-any
+const { Deno } = globalThis as any;
+
+const ALREADY_WARNED_DEPRECATED = new Set<string>();
+
+interface WarnDeprecatedApiConfig {
+  /** The name of the deprecated API. */
+  name: string;
+  /** The stack trace of the deprecated API. */
+  stack: string;
+  /** The version in which the API will be removed. */
+  version: string;
+  /** An optional message to print. */
+  message?: string;
+}
+
+/**
+ * Prints a warning message to the console for the given deprecated API.
+ *
+ * These warnings can be disabled by setting `NO_DEPRECATION_WARNINGS=1`
+ * in the current process.
+ */
+export function warnDeprecatedApi(config: WarnDeprecatedApiConfig) {
+  const variable = "NO_DEPRECATION_WARNINGS";
+  if (
+    Deno?.permissions.querySync({ name: "env", variable }).state ===
+      "granted" && Deno?.env.get(variable) === "1"
+  ) return;
+
+  // Remove the prepending "Error\n" from the stack output.
+  const stack = config.stack.slice(6);
+  const key = config.name + stack;
```

Comment:
Thought - Could you just set a unique `key` on the `WarnDeprecatedApiConfig` type instead of building a unique key from the name & stack on very function call? That way you don't need to do string manipulation and concatenation on every function call (which can potential slow things down).

Reply:
Doing so would require the maintainer to keep track of the unique keys they use. Using `new Error().stack` allows for each location of the deprecated API's use to be reported. Also, having a slight slow-down during the use of a deprecated API is even more reason to migrate to the updated API.

Reply:
Fair enough. That makes logically sense. I do worry about the lag / time for intermediate libraries being updated to the latest stdlib though. I've seen quite a few libraries using very outdated versions  of stdlib already. Hopefully end user / app consumers of these libraries can motivate upstream updates in reasonably amount of time to avoid being too much of a disruption.

Reply:
Yeah, I hope that this change motivates maintainers to use the latest `deno_std` too 🙂